### PR TITLE
Track.{c,h}

### PR DESCRIPTION
**Pull request by alexlib**
This is a pull request for a heavy piece of tracking subroutines: 

```
trackcorr_c_loop  
trackcorr_c_finish  
trackback  
tracking_run
```

with the tests using two cameras + single particle where one could manually and visually verify the tracking procedure (a lot of debug prints are still commented in the code) and a single four-camera four-frame case that we call test_cavity with complex loop structure of adding particles and choosing trajectories.

### CODE REVIEW

**Comment by yosefm** [comment]
It's a long one, just wanted to say that I'm still processing it. 
One comment is that the tracking loop test is a regression test and it tests very little but uses a lot of test data. Can we make actual unit tests where a short trajectory of known geometry is linked from programmatically-generated targets?

**Comment by alexlib** [comment]
This is correct, it's a sort of regression, but it's not comparing with the existing result - I verified manually changing every step (and found several problems). I cannot create a completely virtual test without determination part - one has to create somehow rt_is files that fit some calibration. Of course I can do it for a single particle, single camera, but then we'll miss a lot of other things. 
Actually this part is so heavy looped that I do not even see how to break it into pieces. I tried to create 3 different loops as functions, but it's not a "reused" code, it's just a very long piece of the same loop.

**Comment by yosefm** [comment]
Ok, we'll add a comment explaining this to the code.

File: liboptv/src/track.c [hunk 1]
```diff
@@ -0,0 +1,1409 @@
...
+
+    if ((v0[0] == -v1[0]) && (v0[1] == -v1[1]) && (v0[2] == -v1[2])) {
+        *angle = 200;
+    } else if ((v0[0] == v1[0]) && (v0[1] == v1[1]) && (v0[2] == v1[2])) {
+        *angle = 0; // otherwise it returns NaN
+    } else {
+        *angle = (200./M_PI) * acos(vec_dot(v0, v1) / vec_norm(v0) \
+            / vec_norm(v1));
+    }
+    if (isnan(*angle)){
```
Comment (yosefm) [inline liboptv/src/track.c]: What's all this block?

File: liboptv/src/track.c [hunk 2]
```diff
@@ -0,0 +1,1409 @@
...
+                }
+            }
+
+            p[0]=p1;
+            p[1]=p2;
+            p[2]=p3;
+            p[3]=p4;
+
+            for (j=0; j<4; j++) if ( p[j] != -999 ) {
+                // printf("neighbour %d at %3.2f %3.2f \n",p[j],next[p[j]].x,next[p[j]].y);
```
Comment (yosefm) [inline liboptv/src/track.c]: Remove all the commented-out stuff

File: liboptv/src/track.c [hunk 3]
```diff
@@ -0,0 +1,1409 @@
...
+/* sortwhatfound sourts the list of candidates in foundpix array
+ * Arguments:
+ * foundpix array, item[] which is as long as: num_cam*MAX_CANDS candidates
+ * integer * counter - pointer to the list of the candidates
+ * integer num_cams - number of cameras in the experiment (typically 1-4)
+ * Returns the sorted array foundpix item[] and a pointer *counter to
+ * an integer number of different candidates
+ */
+
+void sortwhatfound (foundpix item[], int *counter, int num_cams)
```
Comment (yosefm) [inline liboptv/src/track.c]: Naming is awful. What kind of things were found? "what" or "stuff" are not good words. same goes to the header comments and to comments inside. This is an opportunity to actually explain what's going on to the next poor soul who needs to modify this code.

Comment (alexlib) [inline liboptv/src/track.c]: see https://github.com/OpenPTV/openptv/pull/121/commits/bb5ab418bc5baa2de622c98db5165e7e2c41d6eb

**Comment by alexlib** [comment]
hope I improved the submission

**Comment by yosefm** [comment]
Can you run this beautify thing again using 4 spaces for tabs?

**Comment by alexlib** [comment]
fixed indentation

**Comment by yosefm** [comment]
The **cal issue is not fully resolved. ``volumedimention`` still receives Calibration[] (an array of calibration objects) instead of **Calibration (an array of pointers). I'm worried that the tests don't complain of it. The signature and the relevant code of ``volumedimention`` (in ``src/multimed.c``) should be updated.

**Comment by alexlib** [comment]
@yosefm please see PR #129

File: liboptv/tests/check_track.c [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+    read_all_calibration(calib, cpar->num_cams);
+    printf("In test_cavity num cams = %d\n",cpar->num_cams);
+    ret = trackcorr_c_init(calib);
+    
+    trackcorr_c_loop (ret, 10002, display, calib);
+    //trackcorr_c_finish(ret, 10002, display);
+    
+    ck_assert_msg(ret->npart == 672,
+                  "Was expecting npart == 672 but found %f \n", ret->npart);
+    ck_assert_msg(ret->nlinks == 99,
```
Comment (alexlib) [inline liboptv/tests/check_track.c]: I try to run now the old PyPTV to see if it gives the same number. as far as I remember this is it for the old 3D-PTV (Tcl/Tk) version that I try to compile as well.

**Comment by yosefm** [comment]
Alex, have you looked at my PR to this branch?

**Comment by alexlib** [comment]
Yes. Have segfaults because of the permissions in copydir, etc. if /res
folder doesn't exist, I cannot test.

Plus the tracking test fails now, need to find the reason.

**Comment by alexlib** [comment]
@yosefm 

this line https://github.com/alexlib/openptv/pull/20/commits/bb01540635318ab7501cb2726f96818e8f50cfd6#diff-7efe340c12ca53823229c1423624bb0dR13 
and the following change for links from 99  to 12 fail the tests. I'm not sure I understand what is the problem. 
I also do not see the bug you mentioned in `Trip and fix a bug related to the unused value change.` - what is the bug?

**Comment by alexlib** [comment]
all tests pass. recommend to merge. 

the only small problem is that one has to know to create `res/` folders manually both in `test_cavity` and in `track' folders under `tests/testing_fodder`

**Comment by yosefm** [comment]
Alex, please see my newest PR to you. I looked over the build problems you had, and I  they are meaningful. They result from reusing test data after the tests overwrite it. I added a commit to fix this, and also streamline the build process a bit. Please test before we decide to merge.

**Comment by yosefm** [comment]
Ok, if there are no further objections by tomorrow, I'll merge.

**Comment by alexlib** [comment]
+1

**Pull request merged**